### PR TITLE
feat: custom invoice expiry for reverse swaps

### LINKED ISSUE #668: custom expiry for reverse swap invoices

### PR DESCRIPTION
**Pull request by michael1011**
Closes #668

### CODE REVIEW

File: lib/api/v2/routers/SwapRouter.ts [hunk 1]
```diff
@@ -1841,6 +1847,13 @@ class SwapRouter extends RouterBase {
 
     checkPreimageHashLength(preimageHash);
 
+    if (
+      invoiceExpiry !== undefined &&
+      !InvoiceExpiryHelper.isValidExpiry(invoiceExpiry)
+    ) {
+      throw 'invalid invoice expiry';
```
Comment (maybeast) [inline lib/api/v2/routers/SwapRouter.ts:1854]: would errorResponse make sense here instead?

File: lib/service/InvoiceExpiryHelper.ts [hunk 2]
```diff
@@ -28,11 +28,9 @@ class InvoiceExpiryHelper {
...
   }
 
-  public getExpiry = (pair: string): number => {
-    return (
-      this.invoiceExpiry.get(pair) || InvoiceExpiryHelper.defaultInvoiceExpiry
-    );
-  };
+  // 43 200 seconds = 12 hours
+  public static isValidExpiry = (expiry: number) =>
+    expiry >= 60 && expiry <= 43_200;
```
Comment (maybeast) [inline lib/service/InvoiceExpiryHelper.ts:33]: could consider making these configurable?

File: lib/service/InvoiceExpiryHelper.ts [hunk 3]
```diff
@@ -28,11 +28,9 @@ class InvoiceExpiryHelper {
     }
   }
 
-  public getExpiry = (pair: string): number => {
-    return (
-      this.invoiceExpiry.get(pair) || InvoiceExpiryHelper.defaultInvoiceExpiry
-    );
-  };
+  // 43 200 seconds = 12 hours
```
Comment (maybeast) [inline lib/service/InvoiceExpiryHelper.ts:31]:
```suggestion
  // min 60 seconds, max 43 200 seconds = 12 hours
```

**Review by maybeast — APPROVED** [review]

File: lib/service/InvoiceExpiryHelper.ts [hunk 4]
```diff
@@ -55,6 +50,20 @@ class InvoiceExpiryHelper {
 
     return invoiceExpiry;
   };
+
+  public isValidExpiry = (pair: string, expiry: number) =>
+    expiry >= InvoiceExpiryHelper.minInvoiceExpiry &&
+    expiry <=
+      (this.invoiceExpiry.get(pair) ||
+        InvoiceExpiryHelper.defaultInvoiceExpiry);
```
Comment (jackstar12) [inline lib/service/InvoiceExpiryHelper.ts:58]: why does it have to be less than default? shouldnt this use a `defaultMaxInvoiceExpiry`?

File: lib/service/InvoiceExpiryHelper.ts [hunk 5]
```diff
@@ -55,6 +50,20 @@ class InvoiceExpiryHelper {
...
+
+  public getExpiry = (pair: string, customExpiry?: number): number => {
```
Comment (jackstar12) [inline lib/service/InvoiceExpiryHelper.ts:60]: I think we could add the `isValidExpiry` call into this one as the caller might forget to do that otherwise

**Pull request merged**